### PR TITLE
AzureMonitor: Add Azure Resource Graph

### PR DESCRIPTION
**Pull request by shuotli**
Goals
- Azure Resource Graph (ARG) is a service in Azure that queries at scale across a given set of subscriptions so you can effectively govern your environment. The goal is to add the ability to query ARG in Grafana.

Concepts
- Resource Manager currently supports queries over basic resource fields, specifically - Resource name, ID, Type, Resource Group, Subscription, and Location.
- You can select multiple subscriptions through the subscription picker
- Endpoint is POST https://management.azure.com/providers/Microsoft.ResourceGraph/resources?api-version=
2018-09-01-preview
- Authorization is using bearer token
- Request body requires 1. subscriptions as an array of strings and 2. query as a string
- Response data is in a row-major column-minor format.
Screenshot 2021-04-22 162050_LI

Open questions
- How is cloud sovereign supported in ARG?

### CODE REVIEW

File: pkg/tsdb/azuremonitor/azuremonitor.go [hunk 1]
```diff
@@ -128,6 +137,11 @@ func (e *AzureMonitorExecutor) DataQuery(ctx context.Context, dsInfo *models.Dat
 		return plugins.DataResponse{}, err
 	}
 
+	argResult, err := argDatasource.executeTimeSeriesQuery(ctx, azureResourceGraphQueries, *tsdbQuery.TimeRange)
```
Comment (mitulmk) [inline pkg/tsdb/azuremonitor/azuremonitor.go:140]: Not all arg queries are time series queries. Does executeTimeSeriesQuery support non-time series queries?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts [hunk 2]
```diff
@@ -11,6 +11,13 @@ export default function createMockQuery(): AzureMonitorQuery {
       workspace: 'e3fe4fde-ad5e-4d60-9974-e2f3562ffdf2',
     },
 
+    azureResourceGraph: {
+      subscriptions: ['abc-123'],
+      query:
+        '//change this example to create your own time series query\n<table name>                                                              //the table to query (e.g. Usage, Heartbeat, Perf)\n| where $__timeFilter(TimeGenerated)                                      //this is a macro used to show the full chart’s time range, choose the datetime column here\n| summarize count() by <group by column>, bin(TimeGenerated, $__interval) //change “group by column” to a column in your table, such as “Computer”. The $__interval macro is used to auto-select the time grain. Can also use 1h, 5m etc.\n| order by TimeGenerated asc',
```
Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:17]: update

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts [hunk 3]
```diff
@@ -11,6 +11,13 @@ export default function createMockQuery(): AzureMonitorQuery {
...
+      resultFormat: 'time_series',
```
Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:18]: this too

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts (hunk 2 above)
Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:17]: Use actual ARG table names

Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:17]: this too

Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:17]: this too

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts [hunk 4]
```diff
@@ -11,6 +11,13 @@ export default function createMockQuery(): AzureMonitorQuery {
       workspace: 'e3fe4fde-ad5e-4d60-9974-e2f3562ffdf2',
     },
 
+    azureResourceGraph: {
+      subscriptions: ['abc-123'],
```
Comment (mitulmk) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/__mocks__/query.ts:15]: use a dummy guid here

File: pkg/tsdb/azuremonitor/azuremonitor.go (hunk 1 above)
Comment (shuotli) [reply pkg/tsdb/azuremonitor/azuremonitor.go:140]: executeTimeSeriesQuery interpolate Macros into KQL  and it takes time range as a parameter.

**Review by joshhunt — COMMENTED** [review]
Looks great, just a few minor things to sort out :)

**Comment by shuotli** [comment]
> Looks great, just a few minor things to sort out :)



> Looks great, just a few minor things to sort out :)

Thank you for your comments Josh!
I'm getting backend linting error that `plugins.DataResponse` has been deprecated for new work. Do you think I could suppress that or I should rewrite part of my code?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      if (!value) {
+        return;
+      }
+
+      onQueryChange({
+        ...query,
+        azureLogAnalytics: {
+          ...query.azureLogAnalytics,
+          resultFormat: value,
+        },
```
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx:28]: Come to think of it, do we even need this format picker? Based on what I observed, most resources don't have a timestamp column and thus maybe we should just have the table view.
I'll think more about this.

**Comment by shuotli** [comment]
Another thing is I realize the panel couldn't handle when the data is non-string (e.g. an object) and returns `no data`. That happens in this case because some of the resource column is object.
Screenshot 2021-04-27 165459

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+	for _, query := range queries {
+		result.Responses[query.RefID] = e.executeQuery(ctx, query, originalQueries, timeRange)
+	}
+
+	return result, nil
+}
+
+func (e *AzureResourceGraphDatasource) buildQueries(queries []plugins.DataSubQuery,
+	timeRange plugins.DataTimeRange) ([]*AzureResourceGraphQuery, error) {
+	azureResourceGraphQueries := []*AzureResourceGraphQuery{}
```
Comment (dsotirakis) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:71]:
```suggestion
	var azureResourceGraphQueries []*AzureResourceGraphQuery
```

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+			Params:       params,
+			Target:       params.Encode(),
+		})
+	}
+
+	return azureResourceGraphQueries, nil
+}
+
+func (e *AzureResourceGraphDatasource) executeQuery(ctx context.Context, query *AzureResourceGraphQuery,
+	queries []plugins.DataSubQuery, timeRange plugins.DataTimeRange) backend.DataResponse {
```
Comment (dsotirakis) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:114]: Unused `queries []plugins.DataSubQuery`?

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+	queryResult := backend.DataResponse{}
+
+	query.Params.Add("api-version", "2018-09-01-preview")
```
Comment (dsotirakis) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:117]: Why is this hardcoded?

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+		strings.Join(query.Model.Get("azureResourceGraph").Get("subscriptions").MustStringArray(), "','"),
+		query.Model.Get("azureResourceGraph").Get("query").MustString())
+	req, err := e.createRequest(ctx, e.dsInfo, reqBody)
+
+	if err != nil {
+		queryResult.Error = err
+		return queryResult
+	}
+
+	req.URL.Path = path.Join(req.URL.Path, "/providers/Microsoft.ResourceGraph/resources")
```
Comment (dsotirakis) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:143]: Make `"/providers/Microsoft.ResourceGraph/resources"` a constant to this file?

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+		azureResourceGraphTarget := queryJSONModel.AzureResourceGraph
+		azlog.Debug("AzureResourceGraph", "target", azureResourceGraphTarget)
+
+		resultFormat := azureResourceGraphTarget.ResultFormat
+		if resultFormat == "" {
+			resultFormat = timeSeries
+		}
+
+		params := url.Values{}
+		rawQuery, err := KqlInterpolate(query, timeRange, azureResourceGraphTarget.Query)
```
Comment (dsotirakis) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:94]: Don't we need a default value here?

**Comment by dsotirakis** [comment]
@shuotli - thanks for your contribution, added some comments!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx]: This makes sense - ARG doesnt really return time series data, so it should always return it as a table?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx [hunk 11]
```diff
@@ -95,9 +97,33 @@ const SubscriptionField: React.FC<SubscriptionFieldProps> = ({
...
+      if (!change) {
+        return;
+      }
+
+      query.azureResourceGraph.subscriptions = change.map((c) => c.value ?? '');
+
+      onQueryChange(query);
+    },
+    [query, onQueryChange]
+  );
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:111]: `subscriptions` should be a top level property on the query object, parallel with `subscription`. Could you update the `AzureMonitorQuery` type to move `subscriptions` there and remove it from `AzureResourceGraphQuery` (and make the appropriate changes to the backend)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx]: Yes and I think we should also remove the format picker for ARG.

[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx]: Sure, go ahead!

[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/ArgQueryEditor/FormatAsField.tsx]: I have removed "time series" from ARG.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:111]: Fixed.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/common.ts [hunk 12]
```diff
@@ -6,6 +6,9 @@ import { AzureMonitorOption } from '../types';
 export const findOption = (options: AzureMonitorOption[], value: string | undefined) =>
   value ? options.find((v) => v.value === value) ?? { value, label: value } : null;
 
+export const findOptions = (options: AzureMonitorOption[], values: string[] | undefined) =>
+  values ? options.filter((option) => values.find((v) => v === option.value)) : [];
```
Comment (dkozar) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/utils/common.ts:10]: This seems to have O(n^2) time complexity.

I'd rather optimize it with a map:

```js
export const findOptions = (options: AzureMonitorOption[], values: string[] = []) => {
  if (values.length === 0) {
    return [];
  }

  const set = values.reduce((accum, item) => {
    accum.add(item);
    
    return accum;
  }, new Set());

  return options.filter((option) => set.has(option.value));
}
```

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go (hunk 10 above)
Comment (shuotli) [reply pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:94]: No we don't. The timestamp filter will be entered by users.

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+	argResponse, err := e.unmarshalResponse(res)
+	if err != nil {
+		return queryResultErrorWithExecuted(err)
+	}
+
+	frame, err := LogTableToFrame(&argResponse.Data)
+	if err != nil {
+		return queryResultErrorWithExecuted(err)
+	}
+
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:180]: Could you also add this? That way the raw query will be displayed in the Query inspector in the Query editor. It's very helpful when debugging a query.
```go
	if (frame.Meta == nil) {
		frame.Meta = &data.FrameMeta{}
	}
	frame.Meta.ExecutedQueryString = req.URL.RawQuery
```

**Review by sunker — COMMENTED** [review]
Looks great! 

For some reason I'm not able to get any data back from the api. This extremely basic query should work right? I know I have resources in this subscription. 
`resources`

**Comment by sunker** [comment]
Two things that we need to have in place before merging:
* template variable support. I suggest just rendering the ArgQueryEditor as variable query field. Let's talk more about this in tomorrow's meeting
* [docs](https://github.com/grafana/grafana/blob/bfaf9e3e0defdbec9f94fd637e5eb5972ce846f4/docs/sources/datasources/azuremonitor.md#L1) must be updated

Question: Now that ARG is added to this datasource, does the user need change policies/permissions or something on their subscription(s) for it to work?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx [hunk 14]
```diff
@@ -95,9 +97,33 @@ const SubscriptionField: React.FC<SubscriptionFieldProps> = ({
...
-  return (
+  return multiSelect ? (
+    <Field label="Subscriptions">
+      <MultiSelect
+        isClearable
+        value={findOptions(subscriptions, query.subscriptions)}
+        inputId="azure-monitor-subscriptions-field"
+        onChange={onSubscriptionsChange}
+        options={options}
+        width={40}
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:123]: Can we use the same width for this as for the service picker? Think it's 38

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/plugin.json [hunk 15]
```diff
@@ -44,6 +44,21 @@
       },
       "headers": [{ "name": "x-ms-app", "content": "Grafana" }]
     },
+    {
```
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/plugin.json:47]: There are currently no other clouds using ARG.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts [hunk 16]
```diff
@@ -39,12 +40,17 @@ export class AzureMonitorQueryCtrl extends QueryCtrl {
...
+    subscriptions: string[];
     azureMonitor: AzureMetricQuery;
     azureLogAnalytics: {
       query: string;
       resultFormat: string;
       workspace: string;
     };
+    azureResourceGraph: {
+      query: string;
+      resultFormat: string;
```
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts:52]: > Looks great!
> 
> For some reason I'm not able to get any data back from the api. This extremely basic query should work right? I know I have resources in this subscription.
> `resources`

You should be able to get data response. What's your repro steps and what did you get as response?

**Review by sarahzinger — COMMENTED** [review]
Just a small suggestion for a default query, also in testing I noticed some errors in the console when I resized the window might be worth digging into a little bit. Thanks for this!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts [hunk 17]
```diff
@@ -100,6 +106,9 @@ export class AzureMonitorQueryCtrl extends QueryCtrl {
           ? this.datasource.azureLogAnalyticsDatasource.defaultOrFirstWorkspace
           : '',
     },
+    azureResourceGraph: {
+      resultFormat: 'table',
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts:110]: Actually, I think it might be better to hold back on something like that and think about what that default query should be.

I agree there should be some kind of prompt to get users started, but I'm not entirely happy with how it is for Log Analytics at the moment. I don't like how that is a syntactically invalid query that generates errors on first load :(

Happy for this to stay empty for the time being while we have a think about it.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts:52]: `Resources | project name, type, location` produces results, but `Resources` or even `Resources | project name, tags` does not. Perhaps this is something to do with the column type being non-string?

I'm currently checking with others about the best practice for handling these values (as you brought up previously https://github.com/grafana/grafana/pull/33293#issuecomment-828037403), but right now I think the right thing to do is to convert it to a JSON string, just like the ARG Explorer in the Azure Portal does.

[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts:52]: Yes the tags are in json. As I requested above, could you convert to string in the panel?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx (hunk 14 above)
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:123]: Is there a way the sub picker adjust its width according to the selected values while maintaining a min width?

**Comment by sunker** [comment]
@shuotli for this PR to be approved, you need to address the changes we've requested.

**Comment by shuotli** [comment]
> @shuotli for this PR to be approved, you need to address the changes we've requested. You also need to write unit tests for azure-resource-graph-datasource.go and for azure_resource_graph_datasource.ts.
> 
> @grafana/docs-squad would you mind taking a look at the docs changes?

I will work on the unit tests.
So far, I have addressed:
1. templating variable support
2. object as value in the response data
3. syntax error in the query
4. documentaion
What else am I missing?

File: docs/sources/datasources/azuremonitor.md [hunk 18]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
 > **Note** In Grafana 8.0, Application Insights and Insights Analytics will be deprecated and made read-only in favor of querying this data through Metrics and Logs. Existing queries will continue to work, but you cannot edit them.
 
 To prepare for this upcoming change, Application Insights queries can now be made in Metrics, under the "microsoft.insights/components" Namespace. Insights Analytics queries cannot be made within Logs with KQL at this time.
+
+## Query Azure Resource Graph service
+
+{{< docs-imagebox img="/img/docs/v80/azuremonitor-azure-resource-graph.png" class="docs-image--no-shadow" caption="Azure Resource Graph Editor" >}}
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:420]: @joshhunt or @sunker, no images should be in version folders. They should be in folders named after the subect, like azure-monitor. Please review the Grafana style guide for instructions about [Images](https://github.com/grafana/grafana/blob/main/contribute/style-guides/documentation-style-guide.md#images).

File: docs/sources/datasources/azuremonitor.md [hunk 19]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
 > **Note** In Grafana 8.0, Application Insights and Insights Analytics will be deprecated and made read-only in favor of querying this data through Metrics and Logs. Existing queries will continue to work, but you cannot edit them.
 
 To prepare for this upcoming change, Application Insights queries can now be made in Metrics, under the "microsoft.insights/components" Namespace. Insights Analytics queries cannot be made within Logs with KQL at this time.
+
+## Query Azure Resource Graph service
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:418]: Add a short paragraph after this heading to provide context about what this service is and when or why you would want to query it.

File: docs/sources/datasources/azuremonitor.md [hunk 20]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
...
+
+{{< docs-imagebox img="/img/docs/v80/azuremonitor-azure-resource-graph.png" class="docs-image--no-shadow" caption="Azure Resource Graph Editor" >}}
+
+Queries are written in the [Kusto Query Language](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/). A Azure Resource Graph (ARG) query is formatted as table data.
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:422]:
```suggestion
Queries are written in the [Kusto Query Language](https://docs.microsoft.com/en-us/azure/data-explorer/kusto/query/). An Azure Resource Graph (ARG) query is formatted as table data.
```

File: docs/sources/datasources/azuremonitor.md [hunk 21]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
...
+
+If your credentials give you access to multiple subscriptions, you can choose multiple subscriptions before entering queries.
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:424]:
```suggestion
If your credentials give you access to multiple subscriptions, then you can choose multiple subscriptions before entering queries.
```

File: docs/sources/datasources/azuremonitor.md [hunk 22]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
...
+
+```kusto
+Resources
+| project name, type, location
+| order by name asc
+```
+
+The query uses `order by` to sort the properties by the `name` property in ascending (asc) order. You can change what property to sort by and the order (`asc` or `desc`). The query uses `project` to show the listed properties in the results. You can add or remove properties.
+
+Filtering for Azure resources with a tag name of `Environment` that have a value of `Internal`. You can change these to any desired tag key and value. The `=~` in the `type` match tells Resource Graph to be case insensitive. You can project by other properties or add/ remove more. The tag key is case sensitive. `Environment` and `environment` will give different results. For example, a query that returns a list of resources with a specified tag value:
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:436]:
```suggestion
Filtering for Azure resources with a tag name of `Environment` that have a value of `Internal`. You can change these to any desired tag key and value. The `=~` in the `type` match tells Resource Graph to be case insensitive. You can project by other properties or add/ remove more.

The tag key is case-sensitive. `Environment` and `environment` will give different results. For example, a query that returns a list of resources with a specified tag value:
```

File: docs/sources/datasources/azuremonitor.md (hunk 18 above)
Comment (oddlittlebird) [reply docs/sources/datasources/azuremonitor.md:420]: Also, please organize this section with subheadings that indicate what each portion is about. Right now, it is hard to follow.

File: docs/sources/datasources/azuremonitor.md [hunk 23]
```diff
@@ -413,3 +414,53 @@ Application Insights and Insights Analytics are two ways to query the same Azure
...
+
+```kusto
+Resources
+| where tags.environment=~'internal'
+| project name
+```
+
+You can also use `summarize` and `count` to define how to group and aggregate the values by property. For example, returning count of healthy, unhealthy and not applicable resources per recommendation.:
```
Comment (oddlittlebird) [inline docs/sources/datasources/azuremonitor.md:444]:
```suggestion
You can also use `summarize` and `count` to define how to group and aggregate the values by property. For example, returning count of healthy, unhealthy, and not applicable resources per recommendation.:
```

**Review by oddlittlebird — DISMISSED** [review]
I have completed a preliminary docs review. The examples are great, thank you! The content would be improved if you added subheadings to indicate what each portion of the content is about.

Developers, please change the name and location of the image as requested, then update the link AND the link format.

**Comment by oddlittlebird** [comment]
I recommend reviewing the [Grafana documentation style guide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/documentation-style-guide.md) and [Grafana markdown style guide](https://github.com/grafana/grafana/blob/main/contribute/style-guides/documentation-markdown-guide.md) for more information about what we want our documentation to look like.

**Comment by oddlittlebird** [comment]
Sorry for the confusion. There are two different groups reviewing the PR.

The developers review the code. According to @sunker, they have some items still to address before they approve it.

The writers (me) review the documentation. I gave you some new feedback that you need to address in the docs before I approve it. Also, I do not approve any PR with code until the developers approve the code.

**Comment by sunker** [comment]
> I will work on the unit tests.
> So far, I have addressed:
> 
> 1. templating variable support
> 2. object as value in the response data
> 3. syntax error in the query
> 4. documentaion
> 
> What else am I missing?

See my comments in this PR ^

**Comment by shuotli** [comment]
> See my comments in this PR ^

Still, could you be more specific? Do you mean supporting all the clouds?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (shuotli) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/query_ctrl.ts:58]: This is fixed.

File: docs/sources/datasources/azuremonitor.md (hunk 18 above)
Comment (sunker) [reply docs/sources/datasources/azuremonitor.md]: Image fixed! Can you please check to see if it looks ok @oddlittlebird?

Comment (oddlittlebird) [reply docs/sources/datasources/azuremonitor.md]: Yep, looks good

Comment (oddlittlebird) [reply docs/sources/datasources/azuremonitor.md]: @shuotli - The image is correct now, but the comment about subheadings still needs to be addressed.

File: pkg/tsdb/azuremonitor/azure-log-analytics-table-frame.go [hunk 24]
```diff
@@ -10,8 +10,8 @@ import (
 	"github.com/grafana/grafana-plugin-sdk-go/data"
```
[the remark this answers is not in the document]
Comment (shuotli) [reply pkg/tsdb/azuremonitor/azure-log-analytics-table-frame.go:10]: Also changed the name from `LogTableToFrame` to `ResponseTableToFrame`.

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 25]
```diff
@@ -0,0 +1,257 @@
...
+				},
+			},
+		}
+		queryResult.Frames = frames
+		return queryResult
+	}
+
+	reqBody := fmt.Sprintf("{'subscriptions':['%s'], 'query':'%s'}",
+		strings.Join(query.Model.Get("subscriptions").MustStringArray(), "','"),
+		query.Model.Get("azureResourceGraph").Get("query").MustString())
```
[the remark this answers is not in the document]
Comment (shuotli) [reply pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:138]: I tried `"subscriptions": fmt.Sprintf("['%s']", strings.Join(query.Model.Get("subscriptions").MustStringArray(), "','")),`, but got `"subscriptions":"['f7152080-b4e8-47ee-9c85-7f1d0e6b72dc','4dc2cd39-7a89-43d8-bebe-8bb501359891']"`. This won't work because `subscriptions` needs to be an array instead of a string. `json.Marshal` converts everything into a string.

[the remark this answers is not in the document]
Comment (sunker) [reply pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go]: It works if you just skip the strings.Join. I pushed a change!

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+			},
+			queryModel: []plugins.DataSubQuery{
+				{
+					DataSource: &models.DataSource{
+						JsonData: simplejson.NewFromAny(map[string]interface{}{}),
+					},
+					Model: simplejson.NewFromAny(map[string]interface{}{
+						"queryType": "Azure Resource Graph",
+						"azureResourceGraph": map[string]interface{}{
+							"query":        "Resources | project name, properties, '$workspace' | order by name asc",
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go:42]: `$workspace` looks like a template variable, not a macro. Template variables are being interpolated in the frontend, not in the backend. Macros has the format `$__[macro-function-name]()`

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+				},
+			},
+			azureResourceGraphQueries: []*AzureResourceGraphQuery{
+				{
+					RefID:        "A",
+					ResultFormat: "table",
+					URL:          "",
+					Model: simplejson.NewFromAny(map[string]interface{}{
+						"azureResourceGraph": map[string]interface{}{
+							"query":        "Resources | project name, properties, '$workspace' | order by name asc",
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go:56]: Isn't the whole idea of this test that something is supposed to be interpolated? Looks like this query is exactly the same as the one above.

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go [hunk 28]
```diff
@@ -0,0 +1,275 @@
...
+			},
+		}
+		queryResult.Frames = frames
+		return queryResult
+	}
+
+	arg := query.Model.Get("azureResourceGraph")
+	reqBody, err := json.Marshal(map[string]interface{}{
+		"subscriptions": query.Model.Get("subscriptions").MustStringArray(),
+		"query": arg.Get("query").MustString(),
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource.go:138]: At the moment, macros doesn't work and that's because you're picking the original query from the frontend, not the string that was interpolated on line 96. I'll fix this...

File: pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+							"resultFormat": timeSeries,
+						},
+					}),
+					Params: url.Values{"query": {""}},
```
Comment (sunker) [inline pkg/tsdb/azuremonitor/azure-resource-graph-datasource_test.go:60]: Isn't the whole idea of this test to verify that macro interpolation works? Now you're not testing that? I'll fix...

**Comment by oddlittlebird** [comment]
Devs, if you need to push doc changes to a later PR so this code can get into beta2, that is fine. Docs should not be a blocker on this PR.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.test.ts [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+
+  beforeEach(() => {
+    ctx.instanceSettings = {
+      url: 'http://azureresourcegraphapi',
+    };
+
+    ctx.ds = new AzureResourceGraphDatasource(ctx.instanceSettings);
+  });
+
+  describe('When applying template variables', () => {
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_resource_graph/azure_resource_graph_datasource.test.ts:32]: I'm not entirely sure what this is testing. The point of the `applyTemplateVariables` function is to expand variables and store it on the query model before sending it to the backend.

**Comment by sunker** [comment]
@shuotli I've made a few changes today. Can you test this and verify it still works as expected? I'd like to merge this to master tomorrow morning GMT time.

**Comment by shuotli** [comment]
> @shuotli I've made a few changes today. Can you test this and verify it still works as expected? I'd like to merge this to master tomorrow morning GMT time.

Thank you for adding the tests. Both tests work well.
Let me know if you need any change or any help with merging this PR from me.

**Review by sunker — APPROVED** [review]
Merging this PR now. The not addressed docs changes will be taken care of in a new PR.
